### ISSUE TITLE
Improve performance when installing multiple related packages

### ISSUE THREAD

**Issue by db4** [conan-io/conan#5175]
One of my projects uses bincrafters' modular boost packages. Nothing special: it just requires `boost_thread/1.66.0@bincrafters/stable` and `boost_date_time/1.66.0@bincrafters/stable`. All necessary binary packages are available at `conan-center`. But the full list of dependencies turns to be quite long and a CI job spends more than half an hour just to download and install them all (not build!) The harware and network link are quite fast so they shouldn't be an issue. Can this be improved/optimized on conan side? I love the idea of the modular boost but this makes it effectively unusable...

**Comment by lasote** [comment]
Could you enable the log to a file (`export CONAN_TRACE_FILE=/tmp/whatever.log`) and share the file after the create? I would like to see where is it taking so long.

**Comment by db4** [comment]
Ok, my first attempt was not successful: 

```
conan@ec49116429d8:~$ conan install boost_thread/1.66.0@bincrafters/stable
...
ERROR: Error 502 downloading file https://api.bintray.com/conan/bincrafters/public-conan/v1/files/bincrafters/boost_preprocessor/1.66.0/stable/0/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/0/conaninfo.txt
Waiting 0 seconds to retry...
ERROR: Error 502 downloading file https://api.bintray.com/conan/bincrafters/public-conan/v1/files/bincrafters/boost_preprocessor/1.66.0/stable/0/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/0/conaninfo.txt
Waiting 0 seconds to retry...
boost_preprocessor/1.66.0@bincrafters/stable: ERROR: Exception while getting package: 5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9
boost_preprocessor/1.66.0@bincrafters/stable: ERROR: Exception: <class 'conans.errors.ConanException'> Error 502 downloading file https://api.bintray.com/conan/bincrafters/public-conan/v1/files/bincrafters/boost_preprocessor/1.66.0/stable/0/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/0/conaninfo.txt. [Remote: bincrafters]
boost_preprocessor/1.66.0@bincrafters/stable: WARN: Trying to remove package folder: /home/conan/.conan/data/boost_preprocessor/1.66.0/bincrafters/stable/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9
ERROR: Error 502 downloading file https://api.bintray.com/conan/bincrafters/public-conan/v1/files/bincrafters/boost_preprocessor/1.66.0/stable/0/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/0/conaninfo.txt. [Remote: bincrafters]
```
But nevertheless here is a log that I have collected:
 
[install.log](https://github.com/conan-io/conan/files/3197226/install.log)

I now think that something may be wrong with accessing `api.bintray.com` from my site. Should it respond to `ping`? I cannot ping it and `traceroute` does not reach it as well. Where is the correct place to report possible `bintray.com` issues?

**Comment by lasote** [comment]
Yeah, is taking so long to contact bintray, almost a second per request. I will report it to the bintray team.

**Comment by db4** [comment]
Even if this a bintray issue maybe it worth to optimize package downloading process? Say download several packages in parallel like `apt-get` does?

**Comment by lasote** [comment]
We have already an open issue for that, we still don't have it.
https://github.com/conan-io/conan/issues/3452

**Comment by Croydon** [comment]
Since this about downloading this is more issue #1012 I believe.

I'm involved in a project right now where our CI jobs take around 45 minutes and almost all the time is Conan downloading (not building) packages, in particular the boost modular packages

If there is a way to speed this up I would be really grateful

**Comment by lasote** [comment]
Thanks @Croydon

**Comment by db4** [comment]
Looks like bintray now works as it should.  Closing the issue.

**Comment by db4** [comment]
@lasote looks like there is something wrong with bintray again (although [bintray status](https://status.bintray.com/) shows no incidents) and loading modular boost for a CI job takes ages.  Can you report it to bintray team again?

**Comment by lasote** [comment]
asking

**Comment by lasote** [comment]
They fixed some issues related to the performance. Is it better? @db4 ?

**Comment by db4** [comment]
@lasote yes, definitely better. Thanks a lot!

**Comment by Croydon** [comment]
I can confirm that the performance optimization is ridiculously good

The one university project I'm involved in right now is getting build on a private GitLab instance and took previously around 40 Minutes, mostly to download Conan packages from Bintray. Now we are at ~8 Minutes. No other element or influence in the build chain changed AFAIK.

Thanks 👍

**Comment by db4** [comment]
@lasote bintray performance has degraded again... Probably Artifactory's remote repositories is a way to go? Or there is another way to cache `conan-center`/`bincrafters` packages locally for CI?

**Comment by Croydon** [comment]
I can confirm that as well. It's back up at the record high times.

**Comment by lasote** [comment]
@db4 yes, a remote repository in Artifactory to cache the packages would solve this issue, but take into account that remotes and virtuals are only working if you have the revisions mechanism enabled.

I'm reporting again to bintray. Thankd

**Comment by db4** [comment]
@lasote 
> yes, a remote repository in Artifactory to cache the packages would solve this issue.

Surprisingly it does not help much. Everything is cached locally but Artifactory 6.10.4 C++ CE still generates significant delays: total modular boost loading time for one CI job is about 20 minutes. Frankly speaking, I expected much better performance for a local server in a Gigabit LAN with no load.

**Comment by lasote** [comment]
There has to be another issue, Artifactory without network delays is very fast. Maybe there is some issue with the remote (trying to contact the remote somehow?). Have you tried to use the packages in a pure local repository to see the difference? Do you see something in the Artifactory logs? Do you have some more details about the number of packages being uploaded/downloaded? It could be even a Conan issue.

**Comment by db4** [comment]
> There has to be another issue, Artifactory without network delays is very fast

OK, I'll try to investigate it. If I find something, I'll open an issue.

**Comment by datalogics-kam** [comment]
I think I'm having a similar issue. Using remote repos in Artifactory for bincrafters and conan-center. Each is in a virtual repo with a local repo in front for things we had to build locally with different settings.

Checking for latest is taking 6+ seconds for each time it calls the bincrafters Bintray, which is adding 7+ minutes to just fetching locally cached packages. Note duration for the bincrafters call below.

```
{"_action": "REST_API_CALL", "duration": 0.2421729564666748, "headers": {"Authorization": "**********", "User-Agent": "Conan/1.18.1 (Python 3.6.5) python-requests/2.20.1", "X-Client-Anonymous-Id": "**********", "X-Client-Id": "kam"}, "method": "GET", "time": 1565729057.598826, "url": "http://artifactory2.dlogics.com:8081/artifactory/api/conan/conan-center-virtual/v2/conans/boost_type_index/1.67.0/bincrafters/testing/latest"}
{"_action": "REST_API_CALL", "duration": 6.677812814712524, "headers": {"Authorization": "**********", "User-Agent": "Conan/1.18.1 (Python 3.6.5) python-requests/2.20.1", "X-Client-Anonymous-Id": "**********", "X-Client-Id": "kam"}, "method": "GET", "time": 1565729064.278498, "url": "http://artifactory2.dlogics.com:8081/artifactory/api/conan/bincrafters-virtual/v2/conans/boost_type_index/1.67.0/bincrafters/testing/latest"}
```

**Comment by memsharded** [comment]
This issue was about closed bintray, it no longer should be an issue, but please create a new ticket if it is.
